Fix `Colorbar` with all zero heatmap

t-bltg opened the pull request:
# Description

Fixes https://github.com/JuliaPlots/Makie.jl/issues/1738.

With pr:
![hm-mid](https://user-images.githubusercontent.com/13423344/158030280-41b1699e-bcff-43a2-9509-c932acbcff51.png)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] Added unit tests for new algorithms, conversion methods, etc.

SimonDanisch commented on src/makielayout/layoutables/colorbar.jl:
```diff
@@ -78,7 +78,8 @@ function layoutable(::Type{<:Colorbar}, fig_or_scene; bbox = nothing, kwargs...)
         end
-        return something(limits, colorrange, (0, 1))
+        lo, hi = lims = something(limits, colorrange, (0, 1))
+        return lo == hi ? (zero(lo), one(hi)) : lims
```
shouldn't this be something like `(hi - 0.5, hi + 0.5)` ?

t-bltg replied:
Oh, yes I forgot about centering, sorry, will fix. I also tried to fix this after in the pipeline (in `src/makielayout/lineaxis.jl`), but without success.

t-bltg replied:
Oddly, when applying `(oftype(lo, lo - .5), oftype(hi, hi + .5))` instead, I've got this

![z_rng](https://user-images.githubusercontent.com/13423344/157909379-d0d88b3f-36b1-4ae3-9558-bde53cd336c5.png)

, which I'm afraid is obviously wrong.

jkrumbiegel replied:
the heatmap also needs the colorrange

jkrumbiegel replied:
what you did is build a colorbar that takes its colormap from the heatmap but sets different limits, without synching those back to the heatmap.

t-bltg replied:
Thanks for pointing that out. Here is the new output:
![hm-mid](https://user-images.githubusercontent.com/13423344/158030173-7ff489c3-adec-4d6a-a764-66b5cc69d1f2.png)

SimonDanisch reviewed (approved):
Sounds good to me! @jkrumbiegel  any concerns?

jkrumbiegel commented:
Maybe the test should check what the limit values in the colorbar are? Not sure if that's accessible actually. And what happens when the heatmap observable changes so there is now a nonempty range, will the colormap adjust correctly? Maybe that should be included in the test. Also, the 0.5 offset might lead to problems for really large float32 singular values, maybe a factor off 1.5 or so could work? Unless zero

t-bltg commented:
> Maybe the test should check what the limit values in the colorbar are

I've added test for `colorrange` and `limits`.

> And what happens when the heatmap observable changes so there is now a nonempty range

I've also added a test changing either `hm.colorrange` or `cb.limits`.

> Also, the 0.5 offset might lead to problems for really large float32 singular values

As long as you have `hi - lo > eps(Float32)`, I think that should suffice to avoid the initial error `ERROR: ArgumentError: range step cannot be zero`.

t-bltg commented:
I've found a way to change the `colorrange` upstream in the pipeline, and updated the PR.
This is cleaner IMO, please comment.

SimonDanisch commented:
Thank you! That looks like a fair solution!
